### PR TITLE
[rllib] Support for complex / variable-length observation spaces

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Support variable length struct spaces as an extra gym space. A max len is required, and we store these variable length observations with padding.

The decoding is handled similar to the batch format of tf.StructuredTensor:

image

TODO:
- [x] Unit test

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/models/list_batch.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            other lists, there will be extra dimensions for the parent list
+            max lens.
+        max_len (int): The max number of items allowed in each list.
+    """
+
+    def __init__(self, value: TensorType, lengths: List[int], max_len: int):
+        self.value = value
+        self.lengths = lengths
+        self.max_len = max_len
+
```
Comment (ericl) [inline rllib/models/list_batch.py:47]: How to implement unbatch()?

**Comment by sven1977** [comment]
@ericl is this ready for review?

**Comment by ericl** [comment]
It needs some tests, I'll add that next week.

**Review by sven1977 — CHANGES_REQUESTED** [review]

File: doc/source/rllib-models.rst [hunk 2]
```diff
@@ -220,6 +220,21 @@ Self-Supervised Model Losses
 
 You can also use the ``custom_loss()`` API to add in self-supervised losses such as VAE reconstruction loss and L2-regularization.
 
+Variable-length / Complex Observation Spaces
+--------------------------------------------
+
+RLlib supports complex and variable-length observation spaces, including ``gym.spaces.Tuple``, ``gym.spaces.Dict``, and ``rllib.models.extra_spaces.Repeated``. The handling of these spaces is transparent to the user. RLlib internally will insert preprocessors to insert padding for repeated elements, flatten complex observations into a fixed-size vector during transit, and unpack the vector into the structured tensor before sending it to the model. The flattened observation is available to the model as ``input_dict["obs_flat"]``, and the unpacked observation as ``input_dict["obs"]``.
```
[the remark this answers is not in the document]
Comment (ericl) [reply doc/source/rllib-models.rst:226]: Do you mean we should move it out of extra_spaces? I think we should stick with extra spaces for now since that's where the other extra ones are.

File: rllib/examples/complex_struct_space.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+parser.add_argument("--torch", action="store_true")
+parser.add_argument("--eager", action="store_true")
+
+# Constraints on the Repeated space.
+MAX_PLAYERS = 4
+MAX_ITEMS = 7
+MAX_EFFECTS = 2
+
+
+class SimpleRPG(gym.Env):
```
[the remark this answers is not in the document]
Comment (ericl) [reply rllib/examples/complex_struct_space.py:38]: Done

File: rllib/examples/complex_struct_space.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            self.player_space, max_len=MAX_PLAYERS)
+
+    def reset(self):
+        return self.observation_space.sample()
+
+    def step(self, action):
+        return self.observation_space.sample(), 1, True, {}
+
+
+class CustomTorchRPGModel(TorchModelV2, nn.Module):
```
[the remark this answers is not in the document]
Comment (ericl) [reply rllib/examples/complex_struct_space.py:77]: Done

File: doc/source/rllib-models.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib-models.rst:226]: Yeah, I meant to move Repeated and Simplex into `utils.spaces.simplex/repeated.py`.
I thought, I did this already on a separate PR a week ago or so.
https://github.com/ray-project/ray/pull/8608

[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib-models.rst:226]: There is only Simplex and it has been moved in this PR already to `utils/spaces/simplex.py`.
https://github.com/ray-project/ray/pull/8608

**Comment by ericl** [comment]
Moved it.

**Review by sven1977 — APPROVED** [review]
:+1: Super cool to have this new Space. Thanks.
